## Content:
### 01. Importing libraries
### 02. Importing Dataframes
### 03. Check the dimensions of the imported dataframes
### 04. Merging Dataframes
### 05. Confirming the result of merge
### 06. Export the newly created dataframe as "ords_prods_merge" in a .pkl format

# 01. Importing libraries 


In [5]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Dataframes

In [7]:
path= r'C:\Users\Bilel\Desktop\data_analyst_Carrer Foundry\Data Immersion\07-12-24 Instacart Basket Analysis'
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

# 03. Check the dimensions of the imported dataframes

In [9]:
df_ords_prods_combined.shape

(32434489, 11)

In [10]:
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

## <span style="color:green"> observation 
The "df_ords_prods_comnined" dataframe  has the same number of columns and rows and the same head as "df_merged_large"

In [12]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [13]:
df_prods = df_prods.drop(columns =['Unnamed: 0'])

In [14]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [15]:
df_prods.shape

(49672, 5)

# 04. Merging Dataframes

There's a key column 'product_id' observed among these two dataframes indicating a merge is possible, So  the suitable way to combine the df_ords_prods_combined dataframe with the df_prods dataframe, despite their different shape, is to merge them using  their shared column "product_id" with a default inner join . 

## <span style="color:green"> observation 
the error problem here is that the column "_merge" already exists in the df_ords_prods_combined dataframe. Therefore, I have to drop it from the dataframe before operating the merging procedure

In [19]:
# Dropping "_merge" column from "df_ords_prods_combined"
df_ords_prods_combined = df_ords_prods_combined.drop(columns =['_merge'])

In [20]:
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  
0          0  
1          0  
2          0  
3          0  
4          0

In [21]:
df_ords_prods_combined.shape

(32434489, 10)

In [22]:
# Merge the updated dataframes
pd.merge(df_ords_prods_combined , df_prods , on = ['product_id'] , indicator = True )


order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32404854   2977660   206209            13                   1   
32404855   2977660   206209            13                   1   
32404856   2977660   206209            13                   1   
32404857   2977660   206209            13                   1   
32404858   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  new_customer  product_id  \
0                         8                     NaN          True         196   
1                         8                     NaN          True       14084   
2                         8                     NaN          True       12427   
3                         8                     NaN          True       26088   
4                         8                     NaN          True       26405   
...                     ...                     ...           ...         ...   
32404854                 12                     7.0         False       14197   
32404855                 12                     7.0         False       38730   
32404856                 12                     7.0         False       31477   
32404857                 12                     7.0         False        6567   
32404858                 12                     7.0         False       22920   

          add_to_cart_order  reordered  \
0                         1          0   
1                         2          0   
2                         3          0   
3                         4          0   
4                         5          0   
...                     ...        ...   
32404854                  5          1   
32404855                  6          0   
32404856                  7          0   
32404857                  8          0   
32404858                  9          0   

                                     product_name  aisle_id  department_id  \
0                                            Soda        77              7   
1         Organic Unsweetened Vanilla Almond Milk        91             16   
2                             Original Beef Jerky        23             19   
3                      Aged White Cheddar Popcorn        23             19   
4                XL Pick-A-Size Paper Towel Rolls        54             17   
...                                           ...       ...            ...   
32404854                             Tomato Paste         9              9   
32404855          Brownie Crunch High Protein Bar         3             19   
32404856    High Protein Bar Chunky Peanut Butter         3             19   
32404857      Chocolate Peanut Butter Protein Bar         3             19   
32404858      Roasted & Salted Shelled Pistachios       117             19   

          prices _merge  
0            9.0   both  
1           12.5   both  
2            4.4   both  
3            4.7   both  
4            1.0   both  
...          ...    ...  
32404854     5.6   both  
32404855     5.9   both  
32404856     4.2   both  
32404857     4.9   both  
32404858     1.7   both  

[32404859 rows x 15 columns]

# 05. Confirming the result of merge

In [24]:
df_ords_prods_merge = pd.merge(df_ords_prods_combined , df_prods , on = ['product_id'] , indicator = True )


In [25]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

## <span style="color:green"> observation 
The resulting dataframe (after the merge) has 32,404,859 rows, and each of those rows have information found in both input data sets, as we used an inner join for the purposes of this project



# 06. Export the newly created dataframe as "ords_prods_merge" in a .pkl format

In [28]:
#Drop unnecessary columns (_merge) before exporting dataframe
df_ords_prods_merge= df_ords_prods_merge.drop(columns =['_merge'])

In [29]:
# Export the newly crated dataframe
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))